File: Day1/solution.kt
import java.io.File

fun input(file: File) = file.readText().split("\n\n").map {
    it.trim().split("\n").map(String::toInt)
}

fun partOne(data: List<List<Int>>) = data.maxOfOrNull(List<Int>::sum)
fun partTwo(data: List<List<Int>>) = data.map(List<Int>::sum).sortedDescending().take(3).sum()

fun main() {
    val data = input(File("input.txt"))
    println("Part one: ${partOne(data)}")
    println("Part two: ${partTwo(data)}")
}

File: Day1/solution.hs
import Data.List.Split (splitOn)
import Data.List (sortBy)

input :: FilePath -> IO [[Int]]
input s = map (map read . lines) . splitOn "\n\n" <$> readFile s

partOne :: [[Int]] -> Int
partOne = foldl max 0 . map sum

partTwo :: [[Int]] -> Int
partTwo = sum . take 3 . sortBy (flip compare) . map sum

main :: IO ()
main = do
  f <- input "input.txt"
  putStrLn $ "Part one: " ++ show (partOne f)
  putStrLn $ "Part two: " ++ show (partTwo f)
